add getting nms_threshold / iou_threshold from RetinaNet

pavel-esir opened the pull request:
Added getting nms_threshold / iou_threshold from original RetinaNet model
#34761

Code:
* [x]  Comments
* [x]  Code style (PEP8)
* [x]  Transformation generates reshape-able IR: N/A -- no new transformations were added
* [x]  Transformation preserves original framework node names: N/A -- because no transformation were added

Validation:
* [x]  Unit tests: N/A
* [x]  Framework operation tests: N/A -- no new operations were added
* [x]  Transformation tests: N/A -- no new transformations were added
* [x]  e2e model test with an update of ./tests/e2e_oss/utils/reshape_utils.py: N/A
* [x]  Model Optimizer IR Reader check

Documentation:
* [x]  Supported frameworks operations list: N/A
* [x]  Supported **public** models list: N/A
* [x]  New operations specification: N/A
* [x]  Guide on how to convert the **public** model: N/A
* [x]  User guide update: N/A

lazarevevgeny commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -246,9 +246,17 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
         detection_output_op = DetectionOutput(graph, match.custom_replacement_desc.custom_attributes)
+        # get nms from original network
+        for node_name in match.matched_nodes_names():
+            if Node(graph, node_name)['op'] == 'NonMaxSuppression':
```
```
node = Node(graph, node_name)
if node.soft_get('op') == 'NonMaxSuppression':
    if node.is_in_port_connected(3):
        iou_threshold = Node(graph, node_name).in_node(3).value
    else:
        iou_threshold = 0.0 # This is the default value for the TF NMS, right?
    break
```

pavel-esir replied:
In TF it's not defined explicitly https://www.tensorflow.org/api_docs/python/tf/raw_ops/NonMaxSuppressionV3 But conventionally default value is 0.5. But I would prefer not set it here by hand since it must've been defined in the original network

lazarevevgeny replied:
Agree

lazarevevgeny commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -246,9 +246,17 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
+            if Node(graph, node_name)['op'] == 'NonMaxSuppression':
+                iou_threshold = Node(graph, node_name).in_node(3).value
+                break
+        if not iou_threshold:
```
if `iou_threshold` is 0.0 then this condition is True. Please, change to `if iou_threshold is not None:`

pavel-esir replied:
Agree.

lazarevevgeny commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -246,9 +246,17 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
                                                                    applied_width_height_regressions_node)
 
         detection_output_op = DetectionOutput(graph, match.custom_replacement_desc.custom_attributes)
+        # get nms from original network
```
add line `iou_threshold = None`

pavel-esir replied:
Done. Please review.

rkazants commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -79,7 +79,7 @@ def append_variances(priors_scale_node: Node, variance: list):
         stride.out_port(0).connect(shape_part_for_tiling.in_port(3))
 
-        concat_value = Const(graph, {'value': np.array([4])}).create_node()
+        concat_value = Const(graph, {'value': int64_array([4])}).create_node()
         shape_concat = Concat(graph, {'name': name + '/shape_for_tiling', 'in_ports_count': 2,
                                       'axis': np.array(0)}).create_node()
```
please use `create_op_with_const_inputs` for creation of `Concat` node

pavel-esir replied:
Agree. But more reasonable is to use `create_op_node_with_second_input`. Done

rkazants commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -79,7 +79,7 @@ def append_variances(priors_scale_node: Node, variance: list):
         stride.out_port(0).connect(shape_part_for_tiling.in_port(3))
 
-        concat_value = Const(graph, {'value': np.array([4])}).create_node()
+        concat_value = Const(graph, {'value': int64_array([4])}).create_node()
         shape_concat = Concat(graph, {'name': name + '/shape_for_tiling', 'in_ports_count': 2,
                                       'axis': np.array(0)}).create_node()
```
np.array(0) -> int64_array(0)

pavel-esir replied:
Done. Please review.

rkazants commented on model-optimizer/extensions/front/tf/RetinaNetFilteredDetectionsReplacement.py:
```diff
@@ -246,9 +246,18 @@ def generate_sub_graph(self, graph: Graph, match: SubgraphMatch):
+            if Node(graph, node_name)['op'] == 'NonMaxSuppression':
+                iou_threshold = Node(graph, node_name).in_node(3).value
+                break
+        if iou_threshold is not None:
+            raise Error('During {} `iou_threshold` was not retrieved from RetinaNet subgraph'.format(self.replacement_id))
```
assert iou_threshold is not None, 'During {} `iou_threshold` was not retrieved from RetinaNet subgraph'.format(self.replacement_id)

lazarevevgeny replied:
Usually asserts are our internal check for consistency. While 'raise Error' is a more user-friendly error that we want to present to the user. So in this particular case I would recommend to leave 'raise Error'

pavel-esir commented:
Changes made: added extractor for TF NMS v2 because RetinaNet in E2E tests has this version and `get_op_nodes(op='NonMaxSuppression')` unexpectedly skipped them.

I added Maxim to review accidentally. Now replaced him to Vladimir since he is involved in adding new version of NMS. @vgavrilo, please take a look to this PR.

rkazants commented on model-optimizer/extensions/front/tf/non_max_suppression_ext.py:
```diff
@@ -21,6 +21,17 @@
+
+    @classmethod
+    def extract(cls, node):
+        attrs = {'sort_result_descending': 1, 'center_point_box': 0, 'output_type': np.int32}
```
I didn't find `center_point_box`. May you mean `box_encoding`?

lazarevevgeny replied:
Great catch about `center_point_box`! It should be `'box_encoding': 'corner'` instead.

Also, please, remove `center_point_box` attribute from the NMS operation constructor.

`np.int32` is used because TF NMS implementation produces tensor with indices of type `int32`: https://github.com/tensorflow/tensorflow/blob/master/tensorflow/core/ops/image_ops.cc#L972.

pavel-esir replied:
@rkazants, thanks for noting this! Renamed, please review.
Code in extractors is identical but despite this this extractor is needed - it renames op name from `NonMaxSuppressionV2` to `NonMaxSuppression` so that I can find this node in replacer by the common name for all NMSs.
